File: gachisa-backend/src/main/java/com/gachisa/payment/dto/PaymentResponse.java
package com.gachisa.payment.dto;

import com.gachisa.payment.entity.Payment;
import com.gachisa.payment.entity.PaymentAttempt;
import com.gachisa.payment.entity.PaymentAttemptStatus;
import com.gachisa.payment.entity.PaymentMethod;
import com.gachisa.payment.entity.PaymentStatus;
import java.time.LocalDateTime;

public record PaymentResponse(
        Long paymentId,
        Long paymentAttemptId,
        Long participationId,
        Long orderId,
        String pgOrderId,
        String pgPaymentKey,
        int amount,
        PaymentStatus paymentStatus,
        PaymentAttemptStatus attemptStatus,
        PaymentMethod paymentMethod,
        int retryCount,
        LocalDateTime nextRetryAt,
        String failureCode,
        String failureMessage,
        LocalDateTime expiresAt,
        LocalDateTime paidAt,
        LocalDateTime refundedAt,
        LocalDateTime createdAt
) {

    public static PaymentResponse from(Payment payment, PaymentAttempt attempt) {
        return from(payment, attempt, null);
    }

    public static PaymentResponse from(Payment payment, PaymentAttempt attempt, Long orderId) {
        return new PaymentResponse(
                payment.getId(),
                attempt == null ? null : attempt.getId(),
                payment.getParticipationId(),
                orderId,
                attempt == null ? null : attempt.getPgOrderId(),
                attempt == null ? null : attempt.getPgPaymentKey(),
                payment.getAmount(),
                payment.getStatus(),
                attempt == null ? null : attempt.getStatus(),
                attempt == null ? null : attempt.getPaymentMethod(),
                attempt == null ? 0 : attempt.getRetryCount(),
                attempt == null ? null : attempt.getNextRetryAt(),
                attempt == null ? null : attempt.getFailureCode(),
                attempt == null ? null : attempt.getFailureMessage(),
                attempt == null ? null : attempt.getExpiresAt(),
                payment.getPaidAt(),
                payment.getRefundedAt(),
                payment.getCreatedAt()
        );
    }
}

File: gachisa-backend/src/main/kotlin/com/gachisa/payment/dto/PaymentResponse.kt
package com.gachisa.payment.dto

import com.gachisa.payment.entity.Payment
import com.gachisa.payment.entity.PaymentAttempt
import com.gachisa.payment.entity.PaymentAttemptStatus
import com.gachisa.payment.entity.PaymentMethod
import com.gachisa.payment.entity.PaymentStatus
import java.time.LocalDateTime

data class PaymentResponse(
    val paymentId: Long,
    val paymentAttemptId: Long?,
    val participationId: Long,
    val orderId: Long?,
    val pgOrderId: String?,
    val pgPaymentKey: String?,
    val amount: Int,
    val paymentStatus: PaymentStatus,
    val attemptStatus: PaymentAttemptStatus?,
    val paymentMethod: PaymentMethod?,
    val retryCount: Int,
    val nextRetryAt: LocalDateTime?,
    val failureCode: String?,
    val failureMessage: String?,
    val expiresAt: LocalDateTime?,
    val paidAt: LocalDateTime?,
    val refundedAt: LocalDateTime?,
    val createdAt: LocalDateTime,
) {
    companion object {
        fun from(payment: Payment, attempt: PaymentAttempt): PaymentResponse =
            from(payment, attempt, null)

        fun from(payment: Payment, attempt: PaymentAttempt?, orderId: Long?): PaymentResponse = PaymentResponse(
            paymentId = payment.id!!,
            paymentAttemptId = attempt?.id,
            participationId = payment.participationId,
            orderId = orderId,
            pgOrderId = attempt?.pgOrderId,
            pgPaymentKey = attempt?.pgPaymentKey,
            amount = payment.amount,
            paymentStatus = payment.status,
            attemptStatus = attempt?.status,
            paymentMethod = attempt?.paymentMethod,
            retryCount = attempt?.retryCount ?: 0,
            nextRetryAt = attempt?.nextRetryAt,
            failureCode = attempt?.failureCode,
            failureMessage = attempt?.failureMessage,
            expiresAt = attempt?.expiresAt,
            paidAt = payment.paidAt,
            refundedAt = payment.refundedAt,
            createdAt = payment.createdAt,
        )
    }
}
